### PR TITLE
FIX Zero density produces NaN in DARE adapter merging

### PR DESCRIPTION
**Pull request by ErenAta16**
## Status

@BenjaminBossan Ready for review. You asked to be pinged when it is, and my account cannot post comments on this repository, so this description edit is the ping.

## The problem

`add_weighted_adapter` documents `density` as:

> Value between 0 and 1. **0 means all values are pruned** and 1 means no values are pruned.

so `density=0` is a supported input, not a misuse. For the two DARE methods it produces a delta of NaN:

```

density=0.0: {'dare_linear': 'NaN',  'dare_ties': 'NaN',  'ties': 'all-zero', 'magnitude_prune': 'all-zero'}

density=0.25: all four finite

density=0.5:  all four finite

density=1.0:  all four finite

```

`random_pruning` rescales the surviving entries to preserve the expected value of the tensor:

https://github.com/huggingface/peft/blob/a5526d27a9d47d1e8264d5e1b1f96c0fdc79464e/src/peft/utils/merge_utils.py#L67-L72

At `density=0` the Bernoulli mask is all zeros, so `pruned_tensor / density` is `0/0`. The NaN then flows through `dare_linear` / `dare_ties` into the merged adapter weights, and nothing raises along the way. `ties` and `magnitude_prune` route through `magnitude_based_pruning`, which takes a top-k and never divides, so they return the all-zero delta the docstring describes. That asymmetry is what makes it easy to miss: two of the four merge methods behave, two silently produce NaN weights.

## Where it came from

#3152, which is a correct fix. Before it the line was:

```python

torch.div(input=pruned_tensor, other=density)   # result discarded

```

`torch.div` is not in-place, so the rescale never happened and `density=0` returned zeros by accident. #3152 assigned the result back, which fixed the real bug it was aimed at (merged weights systematically scaled down by `density`) and turned the accidental no-op into an actual division by zero.

## The change

Skip the rescale when `density` is 0. Everything has been pruned, so there is no expected value left to preserve, and `pruned_tensor` is already the all-zero tensor `density=0` is documented to produce. Densities in `(0, 1]` are untouched, and the rescale still applies there.

`prune()` already brackets the range on both sides, warning at `density >= 1` and raising below 0. This is the endpoint that fell between them.

After:

```

density=0.0: {'dare_linear': 'all-zero', 'dare_ties': 'all-zero', 'ties': 'all-zero', 'magnitude_prune': 'all-zero'}

```

## Tests

Reworked following @BenjaminBossan's review: no separate test module, the tests sit next to the other `add_weighted_adapter` tests in `tests/test_custom_models.py`, and the zero-density case goes through `add_weighted_adapter` the way a user would reach it rather than calling the merge helpers directly. The two tests kept are the two that were asked for.

- `test_add_weighted_adapter_zero_density_prunes_everything`, parametrized over the four density-aware combination types (`ties`, `dare_ties`, `dare_linear`, `magnitude_prune`), asserting no NaN and an all-zero delta on every LoRA layer. Parametrizing rather than testing only the dare types keeps the record that the magnitude-based ones were always correct here, so a later change that breaks them is visible.

- `test_random_pruning_rescales_for_nonzero_density`, still a direct unit test because it asserts on the rescaling factor itself: survivors at `density=0.5` come out at exactly `2.0` and the mean stays at 1. This is the assertion that would have caught the original #3152 no-op, and it is what stops this being "fixed" by dropping the division altogether.

Removed `test_merge_with_nonzero_density_is_finite` and `test_prune_with_zero_density_returns_zeros` as requested.

One judgement call worth flagging: I kept the parametrization on the zero-density test across all four combination types rather than narrowing it to the two that were broken. `ties` and `magnitude_prune` passing is the informative half, since it records that they route through `magnitude_based_pruning` and never divide. If someone later moves them onto the random-pruning path, the test says so. Happy to cut it to the dare types if you would rather keep it minimal.

Test command:

```

pytest tests/test_custom_models.py -k "add_weighted_adapter or pruning"

15 passed, 18112 deselected

```

Red/green, reverting only the guard in `merge_utils.py`:

```

FAILED ...::test_add_weighted_adapter_zero_density_prunes_everything[dare_linear]

FAILED ...::test_add_weighted_adapter_zero_density_prunes_everything[dare_ties]

2 failed, 3 passed

```

and with the guard back, `5 passed`. The two that pass in both directions are `ties` and `magnitude_prune`, which is the asymmetry the fix is about.

`ruff check` and `ruff format --check` clean on both changed files with ruff 0.16.3, matching the `ruff~=0.16.2` pin in `setup.py`.

The branch is now current with `main` (`a9d039ab`), brought up to date by merge rather than rebase, and is zero commits behind. The 15 tests above and the red/green pair below were both re-measured on it rather than carried over from the older base.

## On keeping density=0 allowed

> I would keep allowing `density=0`, since users might expect it.

Agreed, and that is what this does. `density=0` stays a valid input and returns the all-zero delta the docstring describes, for all four methods. Nothing raises, and the earlier suggestion about turning it into a `ValueError` is dropped.

## AI assistance

AI assistance was used for this patch, so noting it explicitly per `.ai/AGENTS.md`.

- Coordination: https://github.com/huggingface/peft/pull/3486#pullrequestreview-4827678812 is the approval to proceed. @BenjaminBossan confirmed the fix itself is fine and asked for the test rework that this update makes. The bug report and discussion are this PR's own thread.

- Tests run and their results are quoted above, including the red/green pair.

- I closed this PR on 1 August without explaining why, which was the wrong way to leave an open review. There was nothing wrong with the review; I was not in a position to follow it through at the time and should have said so in the thread instead of closing it quietly. Reopened with the requested changes made.

One practical note so the silence is not misread: my account cannot post comments on this repository, so I have put this update in the description rather than in a reply. Anything further from me will arrive the same way, as description edits and commits. @yentur, thank you for picking this up and for the note on the thread, that is what brought me back to it.

### CODE REVIEW

File: tests/test_merge_utils.py
Comment (BenjaminBossan) [inline tests/test_merge_utils.py:1]: It's true that we don't have a dedicated spot for testing `add_weighted_adapter`, I would still not create an extra file. Instead, put them after this test:

https://github.com/huggingface/peft/blob/cf08deeeba784f3264dfb39817fb5c519d12e27d/tests/test_custom_models.py#L4044

Moreover, please go through `add_weighted_adapter` instead of invoking the merging functions directly, as this is how users would do it. For the format of the tests, follow the existing ones.

File: tests/test_merge_utils.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    return [tensor, tensor * 2]
+
+
+@pytest.fixture
+def weights():
+    return torch.tensor([0.5, 0.5])
+
+
+@pytest.mark.parametrize("method", sorted(MERGE_METHODS))
+def test_merge_with_zero_density_prunes_everything(method, task_tensors, weights):
```
Comment (BenjaminBossan) [inline tests/test_merge_utils.py]: I think this and `test_random_pruning_rescales_for_nonzero_density` are the only tests that we really need, you can remove the other tests.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for identifying and fixing this bug. I would keep allowing `density=0`, since users might expect it.

The fix itself is fine but the tests need some work, see my comments.

**Comment by BenjaminBossan** [comment]
@ErenAta16 Why did you close the PR?

**Comment by yentur** [comment]
Update: I opened #3530 picking this up, then closed it again. PEFT's agent policy wants a submitter who can carry an AI-assisted patch end-to-end through review and I cannot, so closing it was better than leaving it open.

Nothing about the finding changes. It is yours, it still reproduces on main, and this thread already has @BenjaminBossan's review on it. Worth reopening when you have time.

Sorry for the noise.

**Comment by BenjaminBossan** [comment]
@ErenAta16 Please ping me when the PR is ready for review.

@yentur Thanks for closing your PR to reduce noise.

**Comment by BenjaminBossan** [comment]
Gentle ping @ErenAta16
